lightning/src/ln/channel.rs

```diff
@@ -1435,6 +1437,223 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
 	/// If we can't release a [`ChannelMonitorUpdate`] until some external action completes, we
 	/// store it here and only release it to the `ChannelManager` once it asks for it.
 	blocked_monitor_updates: Vec<PendingChannelMonitorUpdate>,
+	// If we've sent `commtiment_signed` for an interactive transaction construction,
+	// but have not received `tx_signatures` we MUST set `next_funding_txid` to the
+	// txid of that interactive transaction, else we MUST NOT set it.
+	next_funding_txid: Option<Txid>,
+}
+
+pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider {
+	fn context(&self) -> &ChannelContext<SP>;
+
+	fn context_mut(&mut self) -> &mut ChannelContext<SP>;
+
+	fn interactive_tx_constructor_mut(&mut self) -> &mut Option<InteractiveTxConstructor>;
+
+	fn dual_funding_context(&self) -> &DualFundingChannelContext;
+
+	fn dual_funding_context_mut(&mut self) -> &mut DualFundingChannelContext;
+
+	fn is_initiator(&self) -> bool;
+
+	fn begin_interactive_funding_tx_construction<ES: Deref>(
+		&mut self, entropy_source: &ES, holder_node_id: PublicKey,
+	) -> Result<Option<InteractiveTxMessageSend>, APIError>
+	where ES::Target: EntropySource
+	{
+		let mut funding_outputs = Vec::new();
+		if self.is_initiator() {
+			let tx_out = TxOut {
+				value: Amount::from_sat(self.context().get_value_satoshis()),
+				script_pubkey: self.context().get_funding_redeemscript().to_p2wsh(),
+			};
+			funding_outputs.push(
+				if self.dual_funding_context_mut().their_funding_satoshis.unwrap_or(0) > 0 {
+					OutputOwned::Shared(SharedOwnedOutput::new(tx_out, self.dual_funding_context_mut().our_funding_satoshis))
+				} else {
+					OutputOwned::SharedControlFullyOwned(tx_out)
+				}
+			);
+		}
+
+		// if let Ok((tx_constructor, msg)) = 
+		match InteractiveTxConstructor::new(
+			entropy_source, self.context().channel_id(),
+			self.dual_funding_context_mut().funding_feerate_sat_per_1000_weight,
+			holder_node_id, self.context().counterparty_node_id, self.is_initiator(),
+			self.dual_funding_context_mut().funding_tx_locktime,
+			self.dual_funding_context_mut().our_funding_inputs.clone().unwrap_or(vec![]), funding_outputs,
+			Some((self.context().get_funding_redeemscript(), self.context().channel_value_satoshis))
+		) {
+			Ok((tx_constructor, msg)) => {
+				self.interactive_tx_constructor_mut().replace(tx_constructor);
+				Ok(msg)
+			},
+			Err(err) => {
+				Err(APIError::APIMisuseError { err:
+					match err {
+				        AbortReason::MissingFundingOutput => "Funding output missing".into(),
```

Comment:
Nit: Please check your white space settings, there are mixed tabs and spaces used in these lines

Reply:
The settings are fine for my editor, it's when rust-analyzer fills in branches that it defaults to spaces. I'll have a look at changing its config.